tests/contrib/io/matplotlib/test_matplotlib_s3_writer.py

```diff
@@ -0,0 +1,238 @@
+#     or use the QuantumBlack Trademarks in any other manner that might cause
+# confusion in the marketplace, including but not limited to in advertising,
+# on websites, or on software.
+#
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import json
+
+import matplotlib.pyplot as plt
+import pytest
+import s3fs
+from moto import mock_s3
+
+from kedro.contrib.io.matplotlib import MatplotlibS3Writer
+from kedro.io import DataSetError
+
+BUCKET_NAME = "test_bucket"
+AWS_CREDENTIALS = dict(aws_access_key_id="testing", aws_secret_access_key="testing")
+KEY_PATH = "matplotlib"
+COLOUR_LIST = ["blue", "green", "red"]
+
+
+@pytest.fixture
+def mock_single_plot():
+    plt.plot([1, 2, 3], [4, 5, 6])
+    return plt
+
+
+@pytest.fixture
+def mock_list_plot():
+    plots_list = []
+    colour = "red"
+    for index in range(5):  # pylint: disable=unused-variable
+        plots_list.append(plt.figure())
+        plt.plot([1, 2, 3], [4, 5, 6], color=colour)
+    return plots_list
+
+
+@pytest.fixture
+def mock_dict_plot():
+    plots_dict = {}
+    for colour in COLOUR_LIST:
+        plots_dict[colour] = plt.figure()
+        plt.plot([1, 2, 3], [4, 5, 6], color=colour)
+        plt.close()
+    return plots_dict
+
+
+@pytest.fixture
+def mocked_s3_bucket():
+    """Create a bucket for testing using moto."""
+    with mock_s3():
+        conn = s3fs.core.boto3.client("s3", **AWS_CREDENTIALS)
+        conn.create_bucket(Bucket=BUCKET_NAME)
+        yield conn
+
+
+@pytest.fixture
+def mocked_encrypted_s3_bucket():
+    bucket_policy = {
+        "Version": "2012-10-17",
+        "Id": "PutObjPolicy",
+        "Statement": [
+            {
+                "Sid": "DenyUnEncryptedObjectUploads",
+                "Effect": "Deny",
+                "Principal": "*",
+                "Action": "s3:PutObject",
+                "Resource": "arn:aws:s3:::{}/*".format(BUCKET_NAME),
+                "Condition": {"Null": {"s3:x-amz-server-side-encryption": "aws:kms"}},
+            }
+        ],
+    }
+    bucket_policy = json.dumps(bucket_policy)
+
+    with mock_s3():
+        conn = s3fs.core.boto3.client("s3", **AWS_CREDENTIALS)
+        conn.create_bucket(Bucket=BUCKET_NAME)
+        conn.put_bucket_policy(Bucket=BUCKET_NAME, Policy=bucket_policy)
+
+        yield conn
+
+
+@pytest.fixture
+def plot_writer(mocked_s3_bucket):  # pylint: disable=unused-argument
+    def _matplotlibwriters3():
+        return MatplotlibS3Writer(
+            bucket_name=BUCKET_NAME, filepath=KEY_PATH, credentials=AWS_CREDENTIALS
+        )
+
+    return _matplotlibwriters3()
+
+
+def test_save_data(tmp_path, mock_single_plot, plot_writer, mocked_s3_bucket):
+    """Test saving single matplotlib plot to S3."""
+
+    plot_writer.save(mock_single_plot)
+
+    expected_path = tmp_path / "downloaded_image.png"
+    actual_filepath = tmp_path / "locally_saved.png"
+
+    plt.savefig(str(actual_filepath))
```

Comment:
Why do you need to save locally? You can save the plot object to S3, and download it and check? Same for list and dict test.

Reply:
When I started, I was copying the pattern from the local matplotlib tests. But to me the current pattern makes sense, you want to check that you still get the same thing using matplotlib's native save method, vs save to buffer and push to s3 and re-download.